### PR TITLE
[10.x] HotFix: throw captured `UniqueConstraintViolationException` if there are no matching records on `SELECT` retry

### LINKED ISSUE #48235: Unexpected NULL in createOrFirst() with Multiple Unique Constraints
### Laravel Version

10.21.0

### PHP Version

8.2.8

### Database Driver & Version

_No response_

### Summary
The current implementation of createOrFirst() in Laravel's Eloquent Builder has a behavior that could lead to unintended consequences when there are multiple unique constraints.

### Current Behavior
The createOrFirst() method tries to create a record first and, if a UniqueConstraintViolationException occurs, it falls back to finding the record based on $attributes.

Here is the current implementation:
https://github.com/laravel/framework/blob/96b15c7ac382a9adb4a56d40c640e782d669a112/src/Illuminate/Database/Eloquent/Builder.php#L580-L587

### Issue
Assume that we have two columns id_a and id_b, both of which are unique, and id_a is the primary key.

If the database already has the following records:

(id_a, id_b) = [(1, 100), (2, 200), (3, 300), (4, 400)]
And we call createOrFirst(['id_a' => 5], ['id_b' => 400])

The create() call in the try block will fail because of id_b being unique. However, the catch block only looks for a record where id_a = 5, which will return NULL.

### Steps To Reproduce

1. Create a database table with two unique columns, for example id_a and id_b. Make id_a the primary key.
```sql
CREATE TABLE example (
    id_a INT PRIMARY KEY,
    id_b UNIQUE
);

```

2. Insert some initial records into the table.
```sql
INSERT INTO example (id_a, id_b) VALUES (1, 100), (2, 200), (3, 300), (4, 400);
```

3. Use Laravel's Eloquent Builder to call the createOrFirst() method.
```php
ExampleModel::createOrFirst(['id_a' => 5], ['id_b' => 400]);
```

4. Observe that a UniqueConstraintViolationException is thrown for id_b.
5. Check the returned value; it will be NULL because the catch block is only considering id_a for the search.

### PR DESCRIPTION
**Pull request by mpyw**
- Fixes #48235
- Related:
  - #47973
  - #48144 
  - #48160 
  - #48161 
  - #48192  
  - #48213
  - #48214

There is an issue with `createOrFirst` up to version `v10.21.0`. This causes the following problems when `createOrFirst()` returns NULL.

- `firstOrCreate()` also inadvertently returns NULL.
- **`updateOrCreate()` causes an error because it tries to access the `->wasRecentlyCreated` property on NULL.**

The latter is particularly critical, so if possible, we would like you to release it as a hot fix in version `v10.21.1`. Please consider it.

### CODE REVIEW

**Comment by fuwasegu** [comment]
Looks like a good solution

**Comment by mpyw** [comment]
@taylorotwell @driesvints @tonysm Please check it out 😃

**Comment by mpyw** [comment]
Failing tests looks to be irrelevant. Ready to merge.

**Comment by tonysm** [comment]
@mpyw if it got it right, this isn't "fixing" the problem, right? It's more like letting the error throw so the developer knows about it, which I guess it makes sense. I didn't think about this scenario as I was working on this feature. 🤔

**Comment by mpyw** [comment]
@tonysm Thank you for your comment!

Firstly, it's important to clarify that a `QueryException` was correctly thrown for unique constraint violations on the previous `updateOrCreate` implementation. However, with the recent implementation, silently returning `NULL` led to an unintended `Attempt to read property "wasRecentlyCreated" on null` error in `updateOrCreate`, which is a significant issue.

While you mentioned that this change seems to be just aiding the developer's debug process, it is crucial in the context of a framework providing functionality to web applications. The correct behavior here involves the code either **succeeding** or **throwing a `UniqueConstraintViolationException`**, rather than **leading to an unintended null error**.

Furthermore, it is worth noting that the `UniqueConstraintViolationException` is treated like a `LogicException` in some contexts, while there may be cases where we want to treat it like a `RuntimeException`. In practical terms, there might be scenarios where a developer would like to ignore duplicates for certain `$attributes` while using unique constraint errors for runtime handling of other `$values`. For example, in my application development, I encountered a similar scenario during batch data synchronization, where I had to execute specific handling upon catching a `UniqueConstraintViolationException`.

In conclusion, while it may appear that this change merely aids in debugging, it is a critical adjustment that prevents unintended behavior and enables developers to handle unique constraint violations more appropriately in various scenarios.

**Comment by mpyw** [comment]
@tonysm I think I might understand a bit of what you're saying. Are you suggesting that, in the case of two or more independent unique keys (not composite unique keys), if one of them causes a unique constraint error, we should use the attribute that caused the error as a key to retrieve the existing record?

Indeed, there might be useful cases for this, but attempting to address this would make the operation extremely complex and would likely require changing the method signature to accept groups of `$attributes`. I am a colleague of @fuwasegu, the person who raised the issue, and we did not consider such cases. I believe it is acceptable to consider this issue out of scope for the current response.

**Comment by tonysm** [comment]
@mpyw I get it that the unique violation can be used programatically.

I'm trying to think if we better switch the create or first to do a `firstOrFail` instead of `first` and let it throw a `ModelNotFoundException` instead of UniqueViolationException` (as you did in your recent closed PR).

I _think_ this would signal better that the record couldn't be found using the given attributes.

But this wouldn't solve the issue in `updateOrCreate`, at least not in a BC way, since the `ModelNotFoundException` is not a subclass of the `QueryException`.

**Comment by fuwasegu** [comment]
@tonysm 
Thank you for engaging in a discussion to improve the issue I reported!

I don't have any objections to the changes that @mpyw has made in this PR.

The flow here is:

1. Try to create a record using $attributes and $values -> if successful, return it.
2. If a Unique Key error prevents creation, we assume a record already exists and try to fetch and return it.

You argue that since the final operation is a 'SELECT', the appropriate exception to throw would be ModelNotFoundException. I understand that perspective.

However, this behavior contradicts what a developer might expect. They would assume that the retrieval is based on $attributes failing the unique constraint, while in reality, it's $values that's causing the conflict.
It is counterintuitive to search using $attributes and get no result when it appears that $attributes violates the unique key constraint. The real reason for this is that $values violates the unique constraint.

Thus, the root cause of the SELECT resulting in NULL is a Unique Key violation. So, shouldn't developers be dealing with UniqueViolationException instead?

**Comment by trevorgehman** [comment]
> They would assume that the retrieval is based on $attributes failing the unique constraint, while in reality, it's $values that's causing the conflict.
It is counterintuitive to search using $attributes and get no result when it appears that $attributes violates the unique key constraint. The real reason for this is that $values violates the unique constraint.

This is exactly the situation we encountered when our test suite failed updating to 10.21.0. We were getting type errors because `firstOrCreate` suddenly started returning `null`. The reason was we had a unique constraint on a column in the `values` list and were listening for a `QueryException` to check if it was caused by the constraint. But then that exception stopped being thrown.

I think throwing the `UniqueConstraintViolationException` exception is the right choice. Just my $0.02.
